File: app/src/main/java/cn/guluwa/gulumusic/data/remote/RemoteSongsDataSource.java
package cn.guluwa.gulumusic.data.remote;

import android.arch.lifecycle.LiveData;

import java.io.File;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cn.guluwa.gulumusic.data.bean.FreshBean;
import cn.guluwa.gulumusic.data.bean.LocalSongBean;
import cn.guluwa.gulumusic.data.bean.SearchResultSongBean;
import cn.guluwa.gulumusic.data.bean.SongPathBean;
import cn.guluwa.gulumusic.data.bean.SongWordBean;
import cn.guluwa.gulumusic.data.bean.TracksBean;
import cn.guluwa.gulumusic.data.bean.ViewDataBean;
import cn.guluwa.gulumusic.data.local.LocalSongsDataSource;
import cn.guluwa.gulumusic.data.remote.retrofit.ApiService;
import cn.guluwa.gulumusic.data.remote.retrofit.RetrofitFactory;
import cn.guluwa.gulumusic.data.total.SongDataSource;
import cn.guluwa.gulumusic.listener.OnResultListener;
import cn.guluwa.gulumusic.manage.AppManager;
import cn.guluwa.gulumusic.manage.Contacts;
import cn.guluwa.gulumusic.utils.AppUtils;
import io.reactivex.Maybe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by guluwa on 2018/1/12.
 */

public class RemoteSongsDataSource implements SongDataSource {

    public static final RemoteSongsDataSource instance = new RemoteSongsDataSource();

    public static RemoteSongsDataSource getInstance() {
        return instance;
    }

    public RemoteSongsDataSource() {
    }

    /**
     * 查询热门歌曲
     *
     * @return
     */
    @Override
    public LiveData<ViewDataBean<List<TracksBean>>> queryNetCloudHotSong() {
        Map<String, Object> map = new HashMap<>();
        map.put("types", "playlist");
        map.put("id", Contacts.NET_CLOUD_HOT_ID);
        return LiveDataObservableAdapter.fromObservableViewData(
                RetrofitFactory.getRetrofit().createApi(ApiService.class)
                        .obtainNetCloudHot(Contacts.SONG_CALLBACK, map)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .map(playListBean -> {
                            for (int i = 0; i < playListBean.getPlaylist().getTracks().size(); i++) {
                                playListBean.getPlaylist().getTracks().get(i).setSinger(
                                        playListBean.getPlaylist().getTracks().get(i).getAr().size() == 0 ?
                                                null : playListBean.getPlaylist().getTracks().get(i).getAr().get(0));
                                playListBean.getPlaylist().getTracks().get(i).setTag(
                                        playListBean.getPlaylist().getTracks().get(i).getAlia().size() == 0 ?
                                                "" : playListBean.getPlaylist().getTracks().get(i).getAlia().get(0));
                                playListBean.getPlaylist().getTracks().get(i).setSource("netease");
                                playListBean.getPlaylist().getTracks().get(i).setIndex(i);
                            }
                            LocalSongsDataSource.getInstance().addSongs(playListBean.getPlaylist().getTracks());
                            return playListBean.getPlaylist().getTracks();
                        })
                        .observeOn(AndroidSchedulers.mainThread())
        );
    }

    /**
     * 歌曲搜索
     *
     * @param freshBean
     * @return
     */
    public LiveData<ViewDataBean<List<SearchResultSongBean>>> searchSongByKeyWord(FreshBean freshBean) {
        Map<String, Object> map = new HashMap<>();
        map.put("types", "search");
        map.put("count", 20);
        map.put("source", AppManager.getInstance().getSearchPlatform());
        map.put("pages", freshBean.page);
        map.put("name", freshBean.key);
        return LiveDataObservableAdapter.fromObservableViewData(
                RetrofitFactory.getRetrofit().createApi(ApiService.class)
                        .searchSongByKeyWord(Contacts.SONG_CALLBACK, map)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
        );
    }

    /**
     * 查询歌曲路径(首页)
     *
     * @param song
     * @return
     */
    @Override
    public void querySongPath(TracksBean song, OnResultListener<SongPathBean> listener) {
        Map<String, Object> map = new HashMap<>();
        map.put("types", "url");
        map.put("id", "".equals(song.getUrl_id()) ? song.getId() : song.getUrl_id());
        map.put("source", song.getSource());
        RetrofitFactory.getRetrofit().createApi(ApiService.class)
                .obtainSongPath(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map(songPathBean -> {
                    songPathBean.setSong(song);
                    songPathBean.setId(song.getId());
                    LocalSongsDataSource.getInstance().addSongPath(songPathBean);
                    return songPathBean;
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listener::success, throwable -> {
                    listener.failed(throwable.getMessage());
                });
    }

    /**
     * 查询歌曲歌词(首页)
     *
     * @param song
     * @return
     */
    @Override
    public void querySongWord(TracksBean song, OnResultListener<SongWordBean> listener) {
        Map<String, Object> map = new HashMap<>();
        map.put("types", "lyric");
        map.put("id", "".equals(song.getLyric_id()) ? song.getId() : song.getLyric_id());
        map.put("source", song.getSource());
        RetrofitFactory.getRetrofit().createApi(ApiService.class)
                .obtainSongWord(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map(songWordBean -> {
                    songWordBean.setSong(song);
                    songWordBean.setId(song.getId());
                    LocalSongsDataSource.getInstance().addSongWord(songWordBean);
                    return songWordBean;
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listener::success, throwable -> listener.failed(throwable.getMessage()));
    }

    /**
     * 查询歌曲封面图(搜索)
     *
     * @param song
     * @param listener
     */
    public void querySearchSongPic(TracksBean song, OnResultListener<SongPathBean> listener) {
        Map<String, Object> map = new HashMap<>();
        map.put("types", "pic");
        map.put("id", "".equals(song.getPic_id()) ? song.getId() : song.getPic_id());
        map.put("source", song.getSource());
        RetrofitFactory.getRetrofit().createApi(ApiService.class)
                .obtainSongPath(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map(songPathBean -> {
                    if (LocalSongsDataSource.getInstance().queryLocalSong(song.getId(), song.getName()) == null ) {
                        song.getAl().setPicUrl(songPathBean.getUrl());
                        LocalSongsDataSource.getInstance().addLocalSong(AppUtils.getLocalSongBean(song));
                    } else {
                        System.out.println("歌曲已存在");
                    }
                    return songPathBean;
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listener::success, throwable -> listener.failed(throwable.getMessage()));
    }

    /**
     * 歌曲下载
     *
     * @param songPathBean
     * @param songName
     * @param listener
     */
    public void downloadSongFile(SongPathBean songPathBean, String songName, OnResultListener<File> listener) {
        RetrofitFactory.getRetrofit().createApi(ApiService.class)
                .downloadSongFile(songPathBean.getUrl())
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map(responseBody -> {
                    if (LocalSongsDataSource.getInstance().queryLocalSong(songPathBean.getId(), songPathBean.getSong().getName()) == null) {
                        LocalSongsDataSource.getInstance().addLocalSong(AppUtils.getLocalSongBean(songPathBean.getSong()));
                    } else {
                        System.out.println("歌曲已存在");
                    }
                    return responseBody;
                })
                .observeOn(Schedulers.io())
                .map(responseBody -> AppUtils.writeSong2Disk(responseBody, songName))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listener::success, throwable -> {
                    System.out.println(throwable.getMessage());
                    listener.failed("歌曲下载失败");
                });
    }
}

File: app/src/main/java/cn/guluwa/gulumusic/data/remote/RemoteSongsDataSource.kt
package cn.guluwa.gulumusic.data.remote

import android.arch.lifecycle.LiveData

import java.io.File
import java.util.HashMap

import cn.guluwa.gulumusic.data.bean.FreshBean
import cn.guluwa.gulumusic.data.bean.LocalSongBean
import cn.guluwa.gulumusic.data.bean.SearchResultSongBean
import cn.guluwa.gulumusic.data.bean.SongPathBean
import cn.guluwa.gulumusic.data.bean.SongWordBean
import cn.guluwa.gulumusic.data.bean.TracksBean
import cn.guluwa.gulumusic.data.bean.ViewDataBean
import cn.guluwa.gulumusic.data.local.LocalSongsDataSource
import cn.guluwa.gulumusic.data.remote.LiveDataObservableAdapter.fromObservableViewData
import cn.guluwa.gulumusic.data.remote.retrofit.RetrofitWorker
import cn.guluwa.gulumusic.data.total.SongDataSource
import cn.guluwa.gulumusic.listener.OnResultListener
import cn.guluwa.gulumusic.manage.AppManager
import cn.guluwa.gulumusic.manage.Contacts
import cn.guluwa.gulumusic.utils.AppUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody

/**
 * Created by guluwa on 2018/1/12.
 */

class RemoteSongsDataSource : SongDataSource {

    /**
     * 查询热门歌曲
     *
     * @return
     */
    override fun queryNetCloudHotSong(): LiveData<ViewDataBean<List<TracksBean>>> {
        val map = HashMap<String, String>()
        map["types"] = "playlist"
        map["id"] = Contacts.NET_CLOUD_HOT_ID
        return fromObservableViewData(
                RetrofitWorker.retrofitWorker
                        .obtainNetCloudHot(Contacts.SONG_CALLBACK, map)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .map<List<TracksBean>> { playListBean ->
                            for (i in 0 until playListBean.playlist!!.tracks!!.size) {
                                playListBean.playlist!!.tracks!![i].singer =
                                        if (playListBean.playlist!!.tracks!![i].ar!!.isEmpty())
                                            null
                                        else
                                            playListBean.playlist!!.tracks!![i].ar!![0]
                                playListBean.playlist!!.tracks!![i].tag =
                                        if (playListBean.playlist!!.tracks!![i].alia!!.isEmpty())
                                            ""
                                        else
                                            playListBean.playlist!!.tracks!![i].alia!![0]
                                playListBean.playlist!!.tracks!![i].source = "netease"
                                playListBean.playlist!!.tracks!![i].index = i
                            }
                            LocalSongsDataSource.getInstance().addSongs(playListBean.playlist!!.tracks!!)
                            playListBean.playlist!!.tracks
                        }
                        .observeOn(AndroidSchedulers.mainThread())
        )
    }

    /**
     * 歌曲搜索
     *
     * @param freshBean
     * @return
     */
    fun searchSongByKeyWord(freshBean: FreshBean): LiveData<ViewDataBean<List<SearchResultSongBean>>> {
        val map = HashMap<String, String>()
        map["types"] = "search"
        map["count"] = "20"
        map["source"] = AppManager.getInstance().searchPlatform
        map["pages"] = freshBean.page.toString()
        map["name"] = freshBean.key
        return fromObservableViewData(
                RetrofitWorker.retrofitWorker
                        .searchSongByKeyWord(Contacts.SONG_CALLBACK, map)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .map { searchResultSongBeans ->
                            for (i in searchResultSongBeans.indices) {
                                val queryLocalSong = LocalSongsDataSource.getInstance().queryLocalSong(
                                        searchResultSongBeans[i].id, searchResultSongBeans[i].name)
                                println(queryLocalSong)
                                searchResultSongBeans[i].isDownLoad = queryLocalSong != null
                            }
                            searchResultSongBeans
                        }
                        .observeOn(AndroidSchedulers.mainThread())
        )
    }

    /**
     * 查询歌曲路径(首页)
     *
     * @param song
     * @return
     */
    override fun querySongPath(song: TracksBean, listener: OnResultListener<SongPathBean>) {
        val map = HashMap<String, String>()
        map["types"] = "url"
        map["id"] = if ("" == song.url_id) song.id else song.url_id
        map["source"] = song.source
        RetrofitWorker.retrofitWorker
                .obtainSongPath(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map { songPathBean ->
                    songPathBean.song = song
                    songPathBean.id = song.id
                    LocalSongsDataSource.getInstance().addSongPath(songPathBean)
                    songPathBean
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listener.success(it) }, { listener.failed(it.message!!) })
    }

    /**
     * 查询歌曲歌词(首页)
     *
     * @param song
     * @return
     */
    override fun querySongWord(song: TracksBean, listener: OnResultListener<SongWordBean>) {
        val map = HashMap<String, String>()
        map["types"] = "lyric"
        map["id"] = if ("" == song.lyric_id) song.id else song.lyric_id
        map["source"] = song.source
        RetrofitWorker.retrofitWorker
                .obtainSongWord(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map { songWordBean ->
                    songWordBean.song = song
                    songWordBean.id = song.id
                    LocalSongsDataSource.getInstance().addSongWord(songWordBean)
                    songWordBean
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listener.success(it) }, { listener.failed(it.message!!) })
    }

    /**
     * 查询歌曲封面图(搜索)
     *
     * @param song
     * @param listener
     */
    fun querySearchSongPic(song: TracksBean, listener: OnResultListener<SongPathBean>) {
        val map = HashMap<String, String>()
        map["types"] = "pic"
        map["id"] = if ("" == song.pic_id) song.id else song.pic_id
        map["source"] = song.source
        RetrofitWorker.retrofitWorker
                .obtainSongPath(Contacts.SONG_CALLBACK, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map { songPathBean ->
                    if (LocalSongsDataSource.getInstance().queryLocalSong(song.id, song.name) == null) {
                        song.al!!.picUrl = songPathBean.url
                        LocalSongsDataSource.getInstance().addLocalSong(AppUtils.getLocalSongBean(song))
                    } else {
                        println("歌曲已存在")
                    }
                    songPathBean
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listener.success(it) }, { listener.failed(it.message!!) })
    }

    /**
     * 歌曲下载
     *
     * @param songPathBean
     * @param songName
     * @param listener
     */
    fun downloadSongFile(songPathBean: SongPathBean, songName: String, listener: OnResultListener<File>) {
        RetrofitWorker.retrofitWorker
                .downloadSongFile(songPathBean.url)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .map<ResponseBody> { responseBody ->
                    if (LocalSongsDataSource.getInstance().queryLocalSong(songPathBean.id, songPathBean.song!!.name) == null) {
                        LocalSongsDataSource.getInstance().addLocalSong(AppUtils.getLocalSongBean(songPathBean.song!!))
                    } else {
                        println("歌曲已存在")
                    }
                    responseBody
                }
                .observeOn(Schedulers.io())
                .map { responseBody -> AppUtils.writeSong2Disk(responseBody, songName) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ listener.success(it!!) }, {
                    println(it.message)
                    listener.failed("歌曲下载失败")
                })
    }

    object SingletonHolder {
        //单例（静态内部类）
        val instance = RemoteSongsDataSource()
    }

    companion object {

        fun getInstance()= SingletonHolder.instance
    }
}
